File: app/test/features/order/models/order_models_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:footverse/features/order/models/coupon_preview_request.dart';
import 'package:footverse/features/order/models/coupon_preview_response.dart';
import 'package:footverse/features/order/models/order_detail_response.dart';
import 'package:footverse/features/order/models/order_item_response.dart';
import 'package:footverse/features/order/models/order_status.dart';
import 'package:footverse/features/order/models/order_summary_response.dart';
import 'package:footverse/features/order/models/payment_method.dart';
import 'package:footverse/features/order/models/payment_status.dart';
import 'package:footverse/features/order/models/place_order_request.dart';

Map<String, dynamic> _previewJson() => <String, dynamic>{
  'code': 'SUMMER10',
  'name': 'Summer Sale',
  'subtotal': 3300000.00,
  'discountAmount': 330000.00,
  'shippingFee': 30000.00,
  'total': 3000000.00,
};

Map<String, dynamic> _itemJson() => <String, dynamic>{
  'id': 900,
  'productVariantId': 100,
  'productName': 'Air Zoom Pegasus',
  'productImageUrl': 'https://cdn.example.com/p1.jpg',
  'color': 'Black',
  'size': '42',
  'unitPrice': 1650000.00,
  'quantity': 2,
  'lineTotal': 3300000.00,
};

Map<String, dynamic> _summaryJson() => <String, dynamic>{
  'id': 700,
  'orderCode': 'FV-20250115103000123',
  'status': 'PENDING',
  'paymentStatus': 'UNPAID',
  'total': 3330000.00,
  'itemCount': 2,
  'createdAt': '2025-01-15T10:30:00',
};

Map<String, dynamic> _detailJson() => <String, dynamic>{
  'id': 700,
  'orderCode': 'FV-20250115103000123',
  'status': 'PENDING',
  'paymentMethod': 'COD',
  'paymentStatus': 'UNPAID',
  'subtotal': 3300000.00,
  'discountAmount': 330000.00,
  'shippingFee': 30000.00,
  'total': 3000000.00,
  'couponCode': 'SUMMER10',
  'shippingRecipientName': 'Nguyen Van A',
  'shippingRecipientPhone': '0901234567',
  'shippingProvince': 'Hà Nội',
  'shippingDistrict': 'Cầu Giấy',
  'shippingWard': 'Dịch Vọng',
  'shippingStreetAddress': '123 Xuân Thủy',
  'note': 'Giao giờ hành chính',
  'items': <Map<String, dynamic>>[_itemJson()],
  'createdAt': '2025-01-15T10:30:00',
  'cancelledAt': null,
  'deliveredAt': null,
};

void main() {
  group('CouponPreviewResponse (dto-spec §14)', () {
    test('maps every field from a real backend payload with a coupon', () {
      final preview = CouponPreviewResponse.fromJson(_previewJson());

      expect(preview.code, 'SUMMER10');
      expect(preview.name, 'Summer Sale');
      expect(preview.subtotal, 3300000.0);
      expect(preview.discountAmount, 330000.0);
      expect(preview.shippingFee, 30000.0);
      expect(preview.total, 3000000.0);
    });

    test('maps a preview without a coupon (null code and name)', () {
      final preview = CouponPreviewResponse.fromJson(
        _previewJson()
          ..['code'] = null
          ..['name'] = null
          ..['discountAmount'] = 0,
      );

      expect(preview.code, isNull);
      expect(preview.name, isNull);
      expect(preview.discountAmount, 0.0);
    });

    test('parses money from JSON numbers and quoted decimal strings', () {
      final fromNumber = CouponPreviewResponse.fromJson(
        _previewJson()..['total'] = 3000000,
      );
      final fromString = CouponPreviewResponse.fromJson(
        _previewJson()..['total'] = '3000000.00',
      );

      expect(fromNumber.total, 3000000.0);
      expect(fromString.total, 3000000.0);
    });
  });

  group('OrderItemResponse (dto-spec §15)', () {
    test('maps every field from a real backend payload', () {
      final item = OrderItemResponse.fromJson(_itemJson());

      expect(item.id, 900);
      expect(item.productVariantId, 100);
      expect(item.productName, 'Air Zoom Pegasus');
      expect(item.productImageUrl, 'https://cdn.example.com/p1.jpg');
      expect(item.color, 'Black');
      expect(item.size, '42');
      expect(item.unitPrice, 1650000.0);
      expect(item.quantity, 2);
      expect(item.lineTotal, 3300000.0);
    });

    test('accepts a null productImageUrl', () {
      final item = OrderItemResponse.fromJson(
        _itemJson()..['productImageUrl'] = null,
      );

      expect(item.productImageUrl, isNull);
    });

    test('parses money from a quoted decimal string', () {
      final item = OrderItemResponse.fromJson(
        _itemJson()
          ..['unitPrice'] = '1650000.00'
          ..['lineTotal'] = '3300000.00',
      );

      expect(item.unitPrice, 1650000.0);
      expect(item.lineTotal, 3300000.0);
    });
  });

  group('OrderSummaryResponse (dto-spec §15)', () {
    test('maps every field from a real backend payload', () {
      final summary = OrderSummaryResponse.fromJson(_summaryJson());

      expect(summary.id, 700);
      expect(summary.orderCode, 'FV-20250115103000123');
      expect(summary.status, OrderStatus.pending);
      expect(summary.paymentStatus, PaymentStatus.unpaid);
      expect(summary.total, 3330000.0);
      expect(summary.itemCount, 2);
      expect(summary.createdAt, DateTime.parse('2025-01-15T10:30:00'));
    });

    test('maps every OrderStatus wire value', () {
      const wireToEnum = <String, OrderStatus>{
        'PENDING': OrderStatus.pending,
        'CONFIRMED': OrderStatus.confirmed,
        'SHIPPING': OrderStatus.shipping,
        'DELIVERED': OrderStatus.delivered,
        'CANCELLED': OrderStatus.cancelled,
      };

      wireToEnum.forEach((wire, expected) {
        final summary = OrderSummaryResponse.fromJson(
          _summaryJson()..['status'] = wire,
        );

        expect(summary.status, expected);
      });
    });

    test('maps the PAID payment status', () {
      final summary = OrderSummaryResponse.fromJson(
        _summaryJson()..['paymentStatus'] = 'PAID',
      );

      expect(summary.paymentStatus, PaymentStatus.paid);
    });
  });

  group('OrderDetailResponse (dto-spec §15)', () {
    test('maps every field and nested items from a real backend payload', () {
      final order = OrderDetailResponse.fromJson(_detailJson());

      expect(order.id, 700);
      expect(order.orderCode, 'FV-20250115103000123');
      expect(order.status, OrderStatus.pending);
      expect(order.paymentMethod, PaymentMethod.cod);
      expect(order.paymentStatus, PaymentStatus.unpaid);
      expect(order.subtotal, 3300000.0);
      expect(order.discountAmount, 330000.0);
      expect(order.shippingFee, 30000.0);
      expect(order.total, 3000000.0);
      expect(order.couponCode, 'SUMMER10');
      expect(order.shippingRecipientName, 'Nguyen Van A');
      expect(order.shippingRecipientPhone, '0901234567');
      expect(order.shippingProvince, 'Hà Nội');
      expect(order.shippingDistrict, 'Cầu Giấy');
      expect(order.shippingWard, 'Dịch Vọng');
      expect(order.shippingStreetAddress, '123 Xuân Thủy');
      expect(order.note, 'Giao giờ hành chính');
      expect(order.items, hasLength(1));
      expect(order.items.first.id, 900);
      expect(order.createdAt, DateTime.parse('2025-01-15T10:30:00'));
      expect(order.cancelledAt, isNull);
      expect(order.deliveredAt, isNull);
    });

    test('accepts null couponCode and note', () {
      final order = OrderDetailResponse.fromJson(
        _detailJson()
          ..['couponCode'] = null
          ..['note'] = null,
      );

      expect(order.couponCode, isNull);
      expect(order.note, isNull);
    });

    test('maps cancelledAt and deliveredAt when present', () {
      final order = OrderDetailResponse.fromJson(
        _detailJson()
          ..['status'] = 'CANCELLED'
          ..['cancelledAt'] = '2025-01-16T09:00:00'
          ..['deliveredAt'] = '2025-01-18T14:00:00',
      );

      expect(order.status, OrderStatus.cancelled);
      expect(order.cancelledAt, DateTime.parse('2025-01-16T09:00:00'));
      expect(order.deliveredAt, DateTime.parse('2025-01-18T14:00:00'));
    });
  });

  group('CouponPreviewRequest (dto-spec §14, validation-spec §9)', () {
    test('serializes the code and cart item ids', () {
      const request = CouponPreviewRequest(
        code: 'SUMMER10',
        cartItemIds: <int>[1, 2, 3],
      );

      expect(request.toJson(), <String, dynamic>{
        'code': 'SUMMER10',
        'cartItemIds': <int>[1, 2, 3],
      });
    });

    test('omits the optional code when it is not set', () {
      const request = CouponPreviewRequest(cartItemIds: <int>[1, 2]);

      expect(request.toJson().containsKey('code'), isFalse);
      expect(request.toJson()['cartItemIds'], <int>[1, 2]);
    });

    test('keeps duplicate cart item ids as entered (server validates)', () {
      const request = CouponPreviewRequest(cartItemIds: <int>[1, 1, 2]);

      expect(request.toJson()['cartItemIds'], <int>[1, 1, 2]);
    });
  });

  group('PlaceOrderRequest (dto-spec §15, validation-spec §10)', () {
    test('serializes every write field', () {
      const request = PlaceOrderRequest(
        cartItemIds: <int>[1, 2],
        addressId: 42,
        couponCode: 'SUMMER10',
        note: 'Giao giờ hành chính',
      );

      expect(request.toJson(), <String, dynamic>{
        'cartItemIds': <int>[1, 2],
        'addressId': 42,
        'couponCode': 'SUMMER10',
        'note': 'Giao giờ hành chính',
      });
    });

    test('omits the optional couponCode and note when not set', () {
      const request = PlaceOrderRequest(cartItemIds: <int>[1], addressId: 42);

      final json = request.toJson();
      expect(json.containsKey('couponCode'), isFalse);
      expect(json.containsKey('note'), isFalse);
      expect(json, <String, dynamic>{
        'cartItemIds': <int>[1],
        'addressId': 42,
      });
    });

    test('keeps duplicate cart item ids as entered (server validates)', () {
      const request = PlaceOrderRequest(
        cartItemIds: <int>[1, 1, 2],
        addressId: 42,
      );

      expect(request.toJson()['cartItemIds'], <int>[1, 1, 2]);
    });
  });
}

File: admin/src/features/order/models/order-models.test.ts
import { describe, expect, it } from 'vitest';

import type { AdminOrderDetailResponse } from './admin-order-detail-response';
import type { AdminOrderSummaryResponse } from './admin-order-summary-response';
import type { OrderItemResponse } from './order-item-response';
import type { UpdateOrderStatusRequest } from './update-order-status-request';

// Captured backend payloads (dto-spec §15/§17). Response types deserialize a
// captured payload field-for-field. `customerId` / `customerFullName` /
// `customerEmail` / `customerPhone` are ADMIN-only (Sprint 12) — present only
// on the dedicated Admin*Response types, never on the customer-scoped DTOs
// (which this feature never models).
describe('OrderItemResponse', () => {
  it('deserializes a captured payload field-for-field', () => {
    const payload = {
      id: 1,
      productVariantId: 7,
      productId: 100,
      productName: 'Air Force 1',
      productImageUrl: 'https://example.com/air-force-1.png',
      color: 'Black',
      size: '42',
      unitPrice: 100,
      quantity: 2,
      lineTotal: 200,
    };

    const item: OrderItemResponse = JSON.parse(JSON.stringify(payload));

    expect(item).toEqual(payload);
  });

  it('deserializes a payload where productImageUrl is explicitly null', () => {
    const payload = {
      id: 1,
      productVariantId: 7,
      productId: 100,
      productName: 'Air Force 1',
      productImageUrl: null,
      color: 'Black',
      size: '42',
      unitPrice: 100,
      quantity: 2,
      lineTotal: 200,
    };

    const item: OrderItemResponse = JSON.parse(JSON.stringify(payload));

    expect(item.productImageUrl).toBeNull();
  });
});

describe('AdminOrderSummaryResponse', () => {
  it('deserializes a captured payload field-for-field, including the customer identity', () => {
    const payload = {
      id: 5,
      orderCode: 'FV-ORDER-5',
      status: 'PENDING',
      paymentStatus: 'UNPAID',
      total: 30200,
      itemCount: 5,
      createdAt: '2026-01-01T00:00:00',
      customerId: 42,
      customerFullName: 'Jane Doe',
      customerEmail: 'jane@example.com',
      customerPhone: '0900000001',
    };

    const summary: AdminOrderSummaryResponse = JSON.parse(JSON.stringify(payload));

    expect(summary).toEqual(payload);
  });
});

describe('AdminOrderDetailResponse', () => {
  it('deserializes a captured payload field-for-field, nesting items and carrying the customer identity', () => {
    const payload = {
      id: 9,
      orderCode: 'FV-ORDER-9',
      status: 'PENDING',
      paymentMethod: 'COD',
      paymentStatus: 'UNPAID',
      subtotal: 200,
      discountAmount: 0,
      shippingFee: 30000,
      total: 30200,
      couponCode: 'SAVE10',
      shippingRecipientName: 'Jane',
      shippingRecipientPhone: '0900000000',
      shippingProvince: 'HCM',
      shippingDistrict: 'D1',
      shippingWard: 'W1',
      shippingStreetAddress: '1 Street',
      note: 'Leave at door',
      items: [
        {
          id: 1,
          productVariantId: 7,
          productId: 100,
          productName: 'Air Force 1',
          productImageUrl: 'https://example.com/air-force-1.png',
          color: 'Black',
          size: '42',
          unitPrice: 100,
          quantity: 2,
          lineTotal: 200,
        },
      ],
      createdAt: '2026-01-01T00:00:00',
      cancelledAt: null,
      deliveredAt: '2026-01-05T00:00:00',
      customerId: 42,
      customerFullName: 'Jane Doe',
      customerEmail: 'jane@example.com',
      customerPhone: '0900000001',
    };

    const detail: AdminOrderDetailResponse = JSON.parse(JSON.stringify(payload));

    expect(detail).toEqual(payload);
    expect(detail.items[0]?.color).toBe('Black');
  });

  it('deserializes a payload where couponCode, note, cancelledAt, and deliveredAt are explicitly null', () => {
    const payload = {
      id: 9,
      orderCode: 'FV-ORDER-9',
      status: 'PENDING',
      paymentMethod: 'COD',
      paymentStatus: 'UNPAID',
      subtotal: 200,
      discountAmount: 0,
      shippingFee: 30000,
      total: 30200,
      couponCode: null,
      shippingRecipientName: 'Jane',
      shippingRecipientPhone: '0900000000',
      shippingProvince: 'HCM',
      shippingDistrict: 'D1',
      shippingWard: 'W1',
      shippingStreetAddress: '1 Street',
      note: null,
      items: [],
      createdAt: '2026-01-01T00:00:00',
      cancelledAt: null,
      deliveredAt: null,
      customerId: 42,
      customerFullName: 'Jane Doe',
      customerEmail: 'jane@example.com',
      customerPhone: '0900000001',
    };

    const detail: AdminOrderDetailResponse = JSON.parse(JSON.stringify(payload));

    expect(detail.couponCode).toBeNull();
    expect(detail.note).toBeNull();
    expect(detail.cancelledAt).toBeNull();
    expect(detail.deliveredAt).toBeNull();
  });
});

describe('UpdateOrderStatusRequest', () => {
  it('serializes the target status', () => {
    const request: UpdateOrderStatusRequest = { status: 'CONFIRMED' };

    expect(JSON.parse(JSON.stringify(request))).toEqual({ status: 'CONFIRMED' });
  });
});
